pittapi/course.py

```diff
@@ -17,319 +17,273 @@
 
+    combined_section_numbers: Optional[List[str]] = None
 
 class Section(NamedTuple):
     term: str
     session: str
     section_number: str
     class_number: str
     section_type: str
-    instructor: str
-    room: str
     status: str
-    meetings_dates: Optional[List[Tuple[datetime, datetime]]] = None
-    days: Optional[List[str]] = None
-    times: Optional[List[str]] = None
-
+    instructors: Optional[List[Instructor]] = None
+    meetings: Optional[List[Meeting]] = None
+    details: Optional[SectionDetails] = None
 
 class Course(NamedTuple):
     subject_code: str
     course_number: str
+    course_id: str
     course_title: str
+    course_description: Optional[str] = None
+    credit_range: Optional[Tuple[int]] = None
+    requisites: Optional[str] = None
+    components: List[Component] = None
+    attributes: List[Attribute] = None
     sections: Optional[List[Section]] = None
 
-
 class Subject(NamedTuple):
     subject_code: str
     courses: Dict[str, Course]
-    term: Optional[str] = None
-
-
-def _get_subject_json() -> Generator[Dict, None, None]:
-    text = requests.get(CLASS_SEARCH_URL).text
-    s = re.search(r"(?=subjects\s*:\s).*,", text)
-    text = s.group()[:-1]
-    text = text[text.find(":") + 1 :]
-    data = json.loads(text)
-
-    for code in data:
-        # Filter out subject codes for non main campuses
-        if not any(v["campus"] == "PIT" for k, v in code["campuses"].items()):
-            continue
-        yield code
-
-
-def get_subject_codes() -> List[str]:
-    subjects = []
-    for code in _get_subject_json():
-        subjects.append(code["subject"])
-    return subjects
-
-
-def get_detailed_subject_codes() -> List[SubjectCode]:
-    subjects = []
-    for code in _get_subject_json():
-        subjects.append(
-            SubjectCode(
-                subject_code=code["subject"],
-                description=code["descr"],
-                academic_group=code["acad_groups"]["group0"]["acad_group"],
-            )
+
+def get_subject_courses(subject: str) -> Subject:
+    subject = _validate_subject(subject)
+
+    json_response = _get_subject_courses(subject)
+    return Subject(
+        subject_code=subject,
+        courses={
+            course["catalog_nbr"] : 
+            Course(
+                subject_code=subject,
+                course_number=course["catalog_nbr"],
+                course_id=course["crse_id"],
+                course_title=course["descr"]
+            ) for course in json_response["courses"]
+        }
+    )
+
+def get_course_details(term: Union[str, int], subject: str, course: Union[str, int]) -> Course:
+    term = _validate_term(term)
+    subject = _validate_subject(subject)
+    course = _validate_course(course)
+
+    internal_course_id = _get_course_id(subject, course)
+    json_response = _get_course_info(internal_course_id)["course_details"]
+    json_response_details = _get_course_sections(internal_course_id, term)
+    return Course(
```

Comment:
This is extremely hairy. There's too much logic nested into this return statement. Declare all your variables in the function body and only return
`return Course(subject_code=subject, components=components, ...)`